src/Arcus.WebApi.Security/Authentication/Certificates/CertificateAuthenticationFilter.cs

```diff
@@ -33,71 +38,95 @@ public async Task OnAuthorizationAsync(AuthorizationFilterContext context)
+                    + "Please configure such an instance (ex. in the Startup) of your application");
             }
-            else
+
+            if (TryGetClientCertificateFromRequest(context.HttpContext, logger, out X509Certificate2 clientCertificate))
             {
-                var validator = services.GetService<CertificateAuthenticationValidator>();
-                if (validator == null)
+                bool isCertificateAllowed = await validator.IsCertificateAllowedAsync(clientCertificate, services);
+                if (isCertificateAllowed)
                 {
-                    throw new KeyNotFoundException(
-                        $"No configured {nameof(CertificateAuthenticationValidator)} instance found in the request services container. "
-                        + "Please configure such an instance (ex. in the Startup) of your application");
+                    LogSecurityEvent(logger, LogLevel.Information, "Client certificate in request is considered allowed according to configured validation requirements");
                 }
-
-                bool isCertificateAllowed = await validator.IsCertificateAllowedAsync(clientCertificate, services);
-                if (!isCertificateAllowed)
+                else
                 {
-                    context.Result = new UnauthorizedResult();
+                    LogSecurityEvent(logger, LogLevel.Error, "Client certificate in request is not considered allowed according to the configured validation requirements", HttpStatusCode.Unauthorized);
+                    context.Result = new UnauthorizedObjectResult("Client certificate in request is not allowed");
                 }
             }
+            else
+            {
+                LogSecurityEvent(logger, LogLevel.Error, "No client certificate is specified in the request while this authentication filter requires a certificate to validate on the configured validation requirements", HttpStatusCode.Unauthorized);
+                context.Result = new UnauthorizedObjectResult("No client certificate found in request");
+            }
         }
 
-        private static X509Certificate2 GetOrLoadClientCertificateFromRequest(HttpContext context, ILogger logger)
+        private static bool TryGetClientCertificateFromRequest(HttpContext context, ILogger logger, out X509Certificate2 clientCertificate)
         {
-            if (context.Connection.ClientCertificate is null)
+            if (context.Connection.ClientCertificate != null)
+            {
+                clientCertificate = context.Connection.ClientCertificate;
+                return clientCertificate != null;
+            }
+
+            if (!context.Request.Headers.TryGetValue(HeaderName, out StringValues headerValues))
             {
-                const string headerName = "X-ARR-ClientCert";
+                logger.LogError("Cannot load client certificate because request header {HeaderName} was not found", HeaderName);
```

Comment:
Same here, it's handled nicely so maybe a trace instead?
```suggestion
                logger.LogTrace("Cannot load client certificate because request header {HeaderName} was not found", HeaderName);
```